src/formats/cif.cpp

```diff
@@ -0,0 +1,205 @@
+//
+// Created by danny305 on May 3rd, 2021.
+//
+
+#include <iostream>
+#include <string>
+#include <vector>
+#include <stdexcept>
+#include <fstream>
+#include <filesystem>
+
+#include <gemmi/cif.hpp>
+#include <gemmi/mmcif.hpp>
+
+#define GEMMI_WRITE_IMPLEMENTATION
+#include <gemmi/to_cif.hpp>
+#undef GEMMI_WRITE_IMPLEMENTATION
+
+#include "chargefw2.h"
+#include "cif.h"
+#include "../structures/molecule_set.h"
+#include "../charges.h"
+#include "../config.h"
+
+
+namespace fs = std::filesystem;
+
+static const std::vector<std::string> atom_site_columns{
+        "group_PDB",
+        "auth_asym_id",
+        "auth_seq_id",
+        "pdbx_PDB_ins_code",
+        "auth_comp_id",
+        "auth_atom_id",
+        "label_alt_id",
+        "type_symbol",
+        "Cartn_x",
+        "Cartn_y",
+        "Cartn_z",
+        "pdbx_PDB_model_num",
+    };
+
+
+class MCRA {
+public:
+    MCRA(const int model,
+         const std::string &chain,
+         const std::string &res_num,
+         const std::string &residue,
+         const std::string &atom)
+        : _model(model), _chain(chain), _res_num(res_num), _residue(residue), _atom(atom)
+        {}
+
+    int find_row(gemmi::cif::Table &table, size_t start_idx = 0) const {
+        size_t idx = start_idx, total_rows = table.length();
+
+        assert(idx < total_rows);
+
+        while (idx < total_rows) {
+            if (this->is_row(table[idx])) {
+                return idx;
+            }
+            ++idx;
+        }
+
+        fmt::print(stderr, "In MCRA::find_row(), unable to find Atom(1, {}, {}, {}, {})\n"
+                    "Looping through entire table to double check...",
+                    this->_chain, this->_res_num, this->_residue, this->_atom); 
+
+        idx = 0;
+        for (const auto site : table) {
+            if (this->is_row(site)) {
+                return idx;
+            }
+            ++idx;
+        }
+        return -1;
+    }
+
+    friend std::ostream &operator<<(std::ostream &os, const MCRA &mcra) {
+        os << "Atom(" << mcra._model << " " << mcra._chain << " " << mcra._res_num << " [" << mcra._residue << "] " << mcra._atom << ")";
+        return os;
+    }
+
+private:
+    bool is_row(const gemmi::cif::Table::Row &row) const {
+        int model = std::stoul(row[11]);
+        const std::string &chain = row[1];
+        const std::string &residue = row[4];
+        const std::string &res_num = row[2];
+        const std::string atom = row[5][0] != '"' ? row[5] : row[5].substr(1, row[5].size() - 2);
+
+        if (_model == model) {
+            if (_chain == chain) {
+                if (_res_num == res_num && _residue == residue) {
+                    if (_atom == atom) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    const int _model;
+    const std::string _chain, _residue, _res_num, _atom;
```

Comment:
Please move attributes (and private things) to the top of the class. Also one attribute per line.

Reply:
sure thing.